src/testing/vortex/logged_process.zig

```diff
@@ -0,0 +1,259 @@
+const AtomicState = std.atomic.Value(State);
+
+// Allocated by init
+child: std.process.Child,
+stdin_thread: std.Thread,
+stderr_thread: std.Thread,
+
+// Lifecycle state
+current_state: AtomicState,
+
+pub fn spawn(
+    allocator: std.mem.Allocator,
+    argv: []const []const u8,
+) !*Self {
+    const self = try allocator.create(Self);
+    errdefer allocator.destroy(self);
+
+    self.* = .{
+        .current_state = AtomicState.init(.running),
+        .child = std.process.Child.init(argv, allocator),
+        .stdin_thread = undefined,
+        .stderr_thread = undefined,
+    };
+
+    self.child.stdin_behavior = .Pipe;
+    self.child.stdout_behavior = .Ignore;
+    self.child.stderr_behavior = .Inherit;
+
+    try self.child.spawn();
+
+    errdefer {
+        _ = self.child.kill() catch {};
+    }
+
+    // Zig doesn't have non-blocking version of child.wait, so we use `BrokenPipe`
+    // on writing to child's stdin to detect if a child is dead in a non-blocking
+    // manner. Checks once a second in a separate thread.
+    _ = try std.posix.fcntl(
+        self.child.stdin.?.handle,
+        std.posix.F.SETFL,
+        @as(u32, @bitCast(std.posix.O{ .NONBLOCK = true })),
+    );
+    self.stdin_thread = try std.Thread.spawn(
+        .{},
+        struct {
+            fn poll_broken_pipe(stdin: std.fs.File, process: *Self) void {
+                while (process.state() == .running) {
+                    std.time.sleep(1 * std.time.ns_per_s);
+                    _ = stdin.write(&.{1}) catch |err| {
+                        switch (err) {
+                            error.WouldBlock => {}, // still running
+                            error.BrokenPipe,
+                            error.NotOpenForWriting,
+                            => {
+                                // Only write the state variable in case the process is still
+                                // considered running. If it was actively terminated, we don't
+                                // want to overwrite that state.
+                                _ = process.current_state.cmpxchgStrong(
+                                    .running,
+                                    .completed,
+                                    .seq_cst,
+                                    .seq_cst,
+                                );
+                                break;
+                            },
+                            else => @panic(@errorName(err)),
+                        }
+                    };
+                }
+            }
+        }.poll_broken_pipe,
+        .{ self.child.stdin.?, self },
+    );
+
+    return self;
+}
+
+pub fn destroy(self: *Self, allocator: std.mem.Allocator) void {
+    assert(self.state() == .terminated or self.state() == .completed);
+    allocator.destroy(self);
+}
+
+pub fn state(self: *Self) State {
+    return self.current_state.load(.seq_cst);
+}
+
+pub fn terminate(
+    self: *Self,
+    signal: comptime_int,
```

Comment:
Why pass the signal here? We should only ever support abrupt, ungraceful shutdown, so this always will be a sigkill.